hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/index/RunIndexActionExecutor.java

```diff
@@ -0,0 +1,384 @@
+      HoodieTableMetadataWriter metadataWriter = table.getMetadataWriter(instantTime)
+          .orElseThrow(() -> new HoodieIndexException(String.format("Could not get metadata writer to run index action for instant: %s", instantTime)));
+      // this will only build index upto base instant as generated by the plan, we will be doing catchup later
+      String indexUptoInstant = indexPartitionInfos.get(0).getIndexUptoInstant();
+      LOG.info("Starting Index Building with base instant: " + indexUptoInstant);
+      metadataWriter.buildMetadataPartitions(context, indexPartitionInfos);
+
+      // get remaining instants to catchup
+      List<HoodieInstant> instantsToCatchup = getInstantsToCatchup(indexUptoInstant);
+      LOG.info("Total remaining instants to index: " + instantsToCatchup.size());
+
+      // reconcile with metadata table timeline
+      String metadataBasePath = getMetadataTableBasePath(table.getMetaClient().getBasePath());
+      HoodieTableMetaClient metadataMetaClient = HoodieTableMetaClient.builder().setConf(hadoopConf).setBasePath(metadataBasePath).build();
+      Set<String> metadataCompletedTimestamps = getCompletedArchivedAndActiveInstantsAfter(indexUptoInstant, metadataMetaClient).stream()
+          .map(HoodieInstant::getTimestamp).collect(Collectors.toSet());
+
+      // index catchup for all remaining instants with a timeout
+      currentIndexedInstant = indexUptoInstant;
+      catchupWithInflightWriters(metadataWriter, instantsToCatchup, metadataMetaClient, metadataCompletedTimestamps);
+      // save index commit metadata and update table config
+      List<HoodieIndexPartitionInfo> finalIndexPartitionInfos = indexPartitionInfos.stream()
+          .map(info -> new HoodieIndexPartitionInfo(
+              info.getVersion(),
+              info.getMetadataPartitionPath(),
+              currentIndexedInstant))
+          .collect(Collectors.toList());
+      HoodieIndexCommitMetadata indexCommitMetadata = HoodieIndexCommitMetadata.newBuilder()
+          .setVersion(LATEST_INDEX_COMMIT_METADATA_VERSION).setIndexPartitionInfos(finalIndexPartitionInfos).build();
+      updateTableConfigAndTimeline(indexInstant, finalIndexPartitionInfos, indexCommitMetadata);
+      return Option.of(indexCommitMetadata);
+    } catch (IOException e) {
+      // abort gracefully
+      abort(indexInstant, indexPartitionInfos.stream().map(HoodieIndexPartitionInfo::getMetadataPartitionPath).collect(Collectors.toSet()));
+      throw new HoodieIndexException(String.format("Unable to index instant: %s", indexInstant));
+    }
+  }
+
+  private void abort(HoodieInstant indexInstant, Set<String> requestedPartitions) {
+    Set<String> inflightPartitions = getInflightMetadataPartitions(table.getMetaClient().getTableConfig());
+    Set<String> completedPartitions = getCompletedMetadataPartitions(table.getMetaClient().getTableConfig());
+    // delete metadata partition
+    requestedPartitions.forEach(partition -> {
```

Comment:
Think about what happens if this fails midway

Reply:
there are multipe points of failure here.. to reduce the blast radius we will make changes to the table config first because after this patch we mostly rely on table configs. additionally, we need more cli commands to allow users to recover easily.. tracking in HUDI-3753